source_files/edge/p_setup.cc

```diff
@@ -2184,133 +1912,118 @@ static void LoadUDMFLineDefs(parser_t *psr)
 			int typenum = -1;
 			int tag = 0;
 			const mobjtype_c *objtype;
-
-			// process thing block
-			while (1)
+			for (;;)
 			{
-				if (!GetNextAssign(psr, (uint8_t*)ident, (uint8_t*)val))
-				{
-					uint8_t *lp = psr->line;
-					while (*lp != 0 && *lp != 0x7D)
-						lp++; // find end of line or '}'
-					if (*lp == 0x7D)
-						break; // end of block
-					if (psr->next >= psr->length)
-						break; // end of lump
-
-					continue; // skip line
-				}
-				// process assignment
-				if (epi::case_cmp(ident, "id") == 0)
-				{
-					tag = str2int(val, 0);
-				}
-				else if (epi::case_cmp(ident, "x") == 0)
-				{
-					x = str2float(val, 0.0f);
-				}
-				else if (epi::case_cmp(ident, "y") == 0)
-				{
-					y = str2float(val, 0.0f);
-				}
-				else if (epi::case_cmp(ident, "height") == 0)
-				{
-					z = str2float(val, 0.0f);
-				}
-				else if (epi::case_cmp(ident, "angle") == 0)
-				{
-					int ta = str2int(val, 0);
-					angle = FLOAT_2_ANG((float)ta);
-				}
-				else if (epi::case_cmp(ident, "type") == 0)
-				{
-					typenum = str2int(val, 0);
-				}
-				else if (epi::case_cmp(ident, "skill1") == 0)
-				{
-					options |= MTF_EASY;
-				}
-				else if (epi::case_cmp(ident, "skill2") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_EASY;
-				}
-				else if (epi::case_cmp(ident, "skill3") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_NORMAL;
-				}
-				else if (epi::case_cmp(ident, "skill4") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_HARD;
-				}
-				else if (epi::case_cmp(ident, "skill5") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_HARD;
-				}
-				else if (epi::case_cmp(ident, "ambush") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_AMBUSH;
-				}
-				else if (epi::case_cmp(ident, "single") == 0)
-				{
-					if (str2bool(val))
-						options &= ~MTF_NOT_SINGLE;
-				}
-				else if (epi::case_cmp(ident, "dm") == 0)
-				{
-					if (str2bool(val))
-						options &= ~MTF_NOT_DM;
-				}
-				else if (epi::case_cmp(ident, "coop") == 0)
-				{
-					if (str2bool(val))
-						options &= ~MTF_NOT_COOP;
-				}
-				// MBF flag
-				else if (epi::case_cmp(ident, "friend") == 0)
-				{
-					if (str2bool(val))
-						options |= MTF_FRIEND;
-				}
-			}
+				if (lex.Match("}"))
+					break;
 
+				std::string key;
+				std::string value;
+				epi::token_kind_e block_tok = lex.Next(key);
+
+				if (block_tok == epi::TOK_EOF)
+					I_Error("Malformed TEXTMAP lump: unclosed block\n");
+
+				if (block_tok != epi::TOK_Ident)
+					I_Error("Malformed TEXTMAP lump: missing key\n");
+
+				if (! lex.Match("="))
+					I_Error("Malformed TEXTMAP lump: missing '='\n");
+
+				block_tok = lex.Next(value);
+
+				if (block_tok == epi::TOK_EOF || block_tok == epi::TOK_ERROR || value == "}")
+					I_Error("Malformed TEXTMAP lump: missing value\n");
+
+				if (! lex.Match(";"))
+					I_Error("Malformed TEXTMAP lump: missing ';'\n");
+
+				if (key == "id")
```

Comment:
... and one last else if block comment  :)  Also makes me think should probably generate ids for keys, though that would be a decision best left up to a profiler.

Reply:
String handling and comparison is a common eater of cycles, especially in loops.   I have seen string table classes, where you can feed a string in and get an id number back, based on hashing.  This makes comparisons very fast and also avoids needing to explicitly define id/string pairs.  If we don't have such a utility class, I can look at adding one and can add a tracker item for it.